osu.Game/Overlays/Toolbar/ToolbarUserButton.cs

```diff
@@ -20,59 +22,103 @@ namespace osu.Game.Overlays.Toolbar
 {
     public class ToolbarUserButton : ToolbarOverlayToggleButton
     {
-        private readonly UpdateableAvatar avatar;
+        private UpdateableAvatar avatar = null!;
+
+        private IBindable<APIUser> localUser = null!;
+
+        private LoadingSpinner spinner = null!;
 
-        [Resolved]
-        private IAPIProvider api { get; set; }
+        private SpriteIcon failingIcon = null!;
 
-        private readonly IBindable<APIState> apiState = new Bindable<APIState>();
+        private IBindable<APIState> apiState = null!;
 
         public ToolbarUserButton()
         {
             AutoSizeAxes = Axes.X;
+        }
 
-            DrawableText.Font = OsuFont.GetFont(italics: true);
-
+        [BackgroundDependencyLoader]
+        private void load(OsuColour colours, IAPIProvider api, LoginOverlay? login)
+        {
             Add(new OpaqueBackground { Depth = 1 });
 
-            Flow.Add(avatar = new UpdateableAvatar(isInteractive: false)
+            Flow.Add(new Container
             {
                 Masking = true,
+                CornerRadius = 4,
                 Size = new Vector2(32),
                 Anchor = Anchor.CentreLeft,
                 Origin = Anchor.CentreLeft,
-                CornerRadius = 4,
                 EdgeEffect = new EdgeEffectParameters
                 {
                     Type = EdgeEffectType.Shadow,
                     Radius = 4,
                     Colour = Color4.Black.Opacity(0.1f),
+                },
+                Children = new Drawable[]
+                {
+                    avatar = new UpdateableAvatar(isInteractive: false)
+                    {
+                        RelativeSizeAxes = Axes.Both,
+                    },
+                    spinner = new LoadingLayer(dimBackground: true, withBox: false, blockInput: false)
+                    {
+                        Anchor = Anchor.Centre,
+                        Origin = Anchor.Centre,
+                        RelativeSizeAxes = Axes.Both,
+                    },
+                    failingIcon = new SpriteIcon
+                    {
+                        Anchor = Anchor.Centre,
+                        Origin = Anchor.Centre,
+                        Alpha = 0,
+                        Size = new Vector2(0.3f),
+                        Icon = FontAwesome.Solid.ExclamationTriangle,
+                        RelativeSizeAxes = Axes.Both,
+                        Colour = colours.YellowLight,
+                    },
                 }
             });
-        }
 
-        [BackgroundDependencyLoader(true)]
-        private void load(LoginOverlay login)
-        {
-            apiState.BindTo(api.State);
+            apiState = api.State.GetBoundCopy();
             apiState.BindValueChanged(onlineStateChanged, true);
 
+            localUser = api.LocalUser.GetBoundCopy();
+            localUser.BindValueChanged(userChanged, true);
```

Comment:
I believe the `LocalUser` value is changed off-thread, making the actions performed by this callback unsafe because this isn't scheduled.

Reply:
Would you be against me making this thread safe? `APIAccess` already has a `Scheduler` so it can easily be fixed. Feels like we're going to miss this one day.

Reply:
I think we explicitly decided not to make it thread-safe, for one reason or another. I'm not ready to make calls on that.

Reply:
Will leave it for now.